config.go

```diff
@@ -0,0 +1,224 @@
+// Copyright 2022 Juan Pablo Tosso and the OWASP Coraza contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package coraza
+
+import (
+	"io/fs"
+
+	"github.com/corazawaf/coraza/v3/internal/corazawaf"
+	"github.com/corazawaf/coraza/v3/loggers"
+	"github.com/corazawaf/coraza/v3/types"
+)
+
+type WAFConfig interface {
+	WithRule(rule *corazawaf.Rule) WAFConfig
+	WithRulesFromFile(path string) WAFConfig
+	WithRulesFromString(rules string) WAFConfig
+
+	WithAuditLog(config AuditLogConfig) WAFConfig
+
+	WithContentInjection() WAFConfig
+
+	WithRequestBodyAccess(config RequestBodyConfig) WAFConfig
+	WithResponseBodyAccess(config ResponseBodyConfig) WAFConfig
+
+	WithDebugLogger(logger corazawaf.DebugLogger) WAFConfig
+	WithErrorLogger(logger corazawaf.ErrorLogCallback) WAFConfig
+
+	WithFSRoot(fs fs.FS) WAFConfig
+}
+
+func NewWAFConfig() WAFConfig {
+	return &wafConfig{}
+}
+
+type RequestBodyConfig interface {
+	WithLimit(limit int) RequestBodyConfig
+	WithInMemoryLimit(limit int) RequestBodyConfig
+}
+
+func NewRequestBodyConfig() RequestBodyConfig {
+	return &requestBodyConfig{}
+}
+
+type ResponseBodyConfig interface {
+	WithLimit(limit int) ResponseBodyConfig
+	WithMimeTypes(mimeTypes []string) ResponseBodyConfig
+}
+
+func NewResponseBodyConfig() ResponseBodyConfig {
+	return &responseBodyConfig{}
+}
+
+type AuditLogConfig interface {
+	LogRelevantOnly() AuditLogConfig
+	WithParts(parts types.AuditLogParts) AuditLogConfig
+	WithLogger(logger loggers.LogWriter) AuditLogConfig
+}
+
+func NewAuditLogConfig() AuditLogConfig {
+	return &auditLogConfig{}
+}
+
+type wafRule struct {
+	rule *corazawaf.Rule
+	str  string
+	file string
+}
+
+type wafConfig struct {
+	rules            []wafRule
+	auditLog         *auditLogConfig
+	contentInjection bool
+	requestBody      *requestBodyConfig
+	responseBody     *responseBodyConfig
+	debugLogger      corazawaf.DebugLogger
+	errorLogger      corazawaf.ErrorLogCallback
+	fsRoot           fs.FS
+}
+
+func (c *wafConfig) WithRule(rule *corazawaf.Rule) WAFConfig {
+	ret := c.clone()
+	ret.rules = append(ret.rules, wafRule{rule: rule})
+	return ret
+}
+
+func (c *wafConfig) WithRulesFromFile(path string) WAFConfig {
+	ret := c.clone()
+	ret.rules = append(ret.rules, wafRule{file: path})
+	return ret
+}
+
+func (c *wafConfig) WithRulesFromString(rules string) WAFConfig {
+	ret := c.clone()
+	ret.rules = append(ret.rules, wafRule{str: rules})
+	return ret
+}
+
+func (c *wafConfig) WithAuditLog(config AuditLogConfig) WAFConfig {
+	ret := c.clone()
+	ret.auditLog = config.(*auditLogConfig)
+	return ret
+}
+
+func (c *wafConfig) WithContentInjection() WAFConfig {
+	ret := c.clone()
+	ret.contentInjection = true
+	return ret
+}
+
+func (c *wafConfig) WithRequestBodyAccess(config RequestBodyConfig) WAFConfig {
+	ret := c.clone()
+	ret.requestBody = config.(*requestBodyConfig)
```

Comment:
Since `RequestBodyConfig` is an interface this casting can of course fail because anyone can implement the interface and attempt to do dynamic stuff. I think for the sake of correctness this interface can have a private method so you make sure you are the only one which can implement it.

Reply:
cc @anuraaga